src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,311 @@
+#    Copyright 2018 SAS Project Authors. All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+"""A SAS-TH HTTP Server implementation, based on v1.2 of the SAS-SAS TS.
+
+A SAS-TH test server could be run by creating the object for the class SasTestHarnessServer
+and by invoking the API start()".
+"""
+
+import hashlib
+import inspect
+import json
+import logging
+import os.path
+import random
+import ssl
+import sys
+import threading
+import time
+from datetime import datetime, timedelta
+from urllib import quote
+from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
+
+CERT_FILE = 'certs/server.cert'
+KEY_FILE = 'certs/server.key'
+CA_CERT = 'certs/ca.cert'
+CIPHERS = [
+    'AES128-GCM-SHA256',              # TLS_RSA_WITH_AES_128_GCM_SHA256
+    'AES256-GCM-SHA384',              # TLS_RSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-ECDSA-AES128-GCM-SHA256',  # TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+    'ECDHE-ECDSA-AES256-GCM-SHA384',  # TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-RSA-AES128-GCM-SHA256',    # TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+]
+
+
+class SasTestHarnessServer(threading.Thread):
+  """
+  Test Harness acting as a HTTP Server running in background and serving Pull/GET requests.
+  """
+  def __init__(self, name, base_url, dump_path, cert_file=None, key_file=None):
+    """ This constructor initializes the SasTestHarnessServer class with required information
+    passed as args
+
+    Args:
+        name: A name for SAS-TH.If multiple SAS-THs are created then this value should be unique.
+        base_url: base url of SAS-TH that the SAS UUT should connect to for FAD records.
+        dump_path: file_path where the FAD records served by the SAS-TH will be placed.
+        cert_file: The certificate used by SAS-TH to authorize with SAS UUT.
+        key_file:  The private key file for the cert_file.
+    """
+    super(SasTestHarnessServer, self).__init__()
+    self.name = name
+    self.http_server_url = base_url
+    self.dump_path = os.path.join(name, dump_path)
+    self.cert_file = cert_file if cert_file is not None else CERT_FILE
+    self.key_file = key_file if key_file is not None else KEY_FILE
+    self.daemon = True
```

Comment:
This   flag is set to make this thread as a daemon thread. This will enable the SAS   Test harness to be automatically terminated when the main test case execution   ends. This has been corrected to call the setDaemon method of the Thread   class to set the correct variable.